protocols/imap/src/main/java/org/apache/james/imap/api/message/UidRange.java

```diff
@@ -38,18 +38,56 @@
             return ranges;
         }
 
+        RangeSet<MessageUid> rangeSet = createSortedRangeSet(ranges);
+        List<Range<MessageUid>> mergedRanges = mergeContiguousRanges(rangeSet);
+        return toUidRanges(mergedRanges);
+    }
+
+    private static RangeSet<MessageUid> createSortedRangeSet(List<UidRange> ranges) {
         RangeSet<MessageUid> rangeSet = TreeRangeSet.create();
         for (UidRange range: ranges) {
             rangeSet.add(Range.closed(range.getLowVal(), range.getHighVal()));
         }
-        List<UidRange> result = new ArrayList<UidRange>();
+        return rangeSet;
+    }
+
+    private static LinkedList<Range<MessageUid>> mergeContiguousRanges(RangeSet<MessageUid> rangeSet) {
+        LinkedList<Range<MessageUid>> mergedRanges = new LinkedList<Range<MessageUid>>();
+        
         for (Range<MessageUid> range: rangeSet.asRanges()) {
```

Comment:
Can you add a test with unordered ranges, to be sure that `TreeRangeSet` do the job:
[[1,2], [10,22], [3,4]] -> [[1,4], [10,22]]